### PR TITLE
automate peerpods image upload & use that in Terraform configs

### PR DESCRIPTION
**Pull request by Freax13**
This PR adds a script/just target for automating the image upload of peerpod VM images and adjusts our Terraform configs to use the image id of the uploaded image.

### CODE REVIEW

File: packages/uplosi.conf.template
Comment (katexochen) [inline packages/uplosi.conf.template:1]: I don't think packages is the right place for a template. Maybe we can just inline it into the script?

File: packages/uplosi.conf.template
Comment (katexochen) [inline packages/uplosi.conf.template:2]: We need a mechanism to update the version on each upload. uplosi has separate image file and -i flag for that.

File: packages/uplosi.conf.template
Comment (katexochen) [inline packages/uplosi.conf.template:11]: Did you check this works with private images? Otherwise public is acceptable, too.

File: flake.nix
Comment (katexochen) [inline flake.nix:1]: flake in uplosi is poorly maintained, just add an override to /overlays with the new version.

File: justfile
Comment (katexochen) [inline justfile:192]: This should be decoupled from create so we don't need to recreate the cluster to iterate on the image.

File: justfile
Comment (katexochen) [inline justfile:195]: Maybe we can add a suffix here so this won't collide with the rg used for AKS?

File: justfile
Comment (katexochen) [inline justfile:196]: Not sure tbh, will uplosi create the rg the image should be uploaded to? Otherwise we need to create that for the dev at some point.

**Review by katexochen — CHANGES_REQUESTED** [review]

File: packages/uplosi.conf.template
Comment (Freax13) [reply packages/uplosi.conf.template:11]: Yes, I did, I created a cluster with this, and worked. I ran into some error when I tried to create a public gallery, but didn't look much into it.

File: infra/azure-peerpods/vars.tf
Comment (Freax13) [reply infra/azure-peerpods/vars.tf:13]: I don't know much about this, but everything seems to work i.e. AFAICT the AKS cluster has all the right permissions to use the image to create VMs. Is that not (yet) expected?

File: justfile
Comment (Freax13) [reply justfile:192]: How about we also add the `just upload-image` and `terraform apply` steps right before the `nix run -L .#scripts.deploy-caa` step in node-installer?

We need this step here because Terraform won't work with the image_id variable missing.
Creating the `just.auto.tfvars` config is idempotent and `terraform apply` is as well. I don't know what `terraform init` is doing.

File: infra/azure-peerpods/vars.tf
Comment (katexochen) [reply infra/azure-peerpods/vars.tf:13]: So you create a rg with a private image and AKS can just access it?

File: justfile
Comment (Freax13) [reply justfile:195]: We already add a suffix in our Terraform configs, so that'd be a bit redundant:
https://github.com/edgelesssys/contrast/blob/0662a2ee4b80bf22441b5a828d2e0f3a6b598320/infra/azure-peerpods/main.tf#L36
Do you still want me to add a suffix?

File: justfile
Comment (Freax13) [reply justfile:196]: No, right, it won't, I forgot about that. I'll fix that.

File: packages/uplosi.conf.template
Comment (katexochen) [reply packages/uplosi.conf.template:2]: I think we need a mechanism to detect whether an image upload is needed or not. I thought we'd to this via version, but there might be other optioins.

File: justfile
Comment (katexochen) [reply justfile:192]: Depends on how well we can detect whether we need to upload or not.

File: infra/azure-peerpods/vars.tf
Comment (Freax13) [reply infra/azure-peerpods/vars.tf:13]: Unless I messed up somehow and AKS is not actually using the private image, everything seems to work.

I used the following steps to test this;
1. `just create AKS-PEER-SNP`.
2. `az aks get-credentials ...`
3. `just node-installer AKS-PEER-SNP`
4. `kubectl create namespace openssl` + `kubectl apply -f workspace/deployment/` where `workspace/deployment` contains the openssl deployment (before `generate`) in which I manually replaced the runtimeClass with `kata-remote`.

File: packages/uplosi.conf.template
Comment (Freax13) [reply packages/uplosi.conf.template:2]: I suppose we could create a file with a name based on `basename $(readlink result)` (e.g. `ay580z91d48zd0irvzfzjph8k4q07sd4-image-podvm-gpu-1-rc1`) and place it `workspace.cache` after uploading. On subsequent runs, we could check if the file already exists and skip the upload.

Comment (Freax13) [reply packages/uplosi.conf.template:2]: We could put the image id of the uploaded image in that file.

File: infra/azure-peerpods/vars.tf
Comment (katexochen) [reply infra/azure-peerpods/vars.tf:13]: Sounds good. There is `just get-credentials-peerpod` :wink:

Comment (Freax13) [reply infra/azure-peerpods/vars.tf:13]: I'm not sure why, but I just got this:
```
  Warning  FailedCreatePodSandBox  16s  kubelet  Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to cre
ate shim task: remote hypervisor call failed: rpc error: code = Unknown desc = creating an instance : Creating instance (<nil>): beginning VM creation or updat
e: PUT https://management.azure.com/subscriptions/0d202bbb-4fa7-4af8-8125-58c269a05435/resourceGroups/tomcontrast_caa_cluster/providers/Microsoft.Compute/virtu
alMachines/podvm-openssl-backend-64f988dbd4-b562v-6e53dddd                                                                                                     
--------------------------------------------------------------------------------                                                                               
RESPONSE 403: 403 Forbidden                                                                                                                                    
ERROR CODE: LinkedAuthorizationFailed                                                                                                                          
--------------------------------------------------------------------------------                                                                               
{                                                                                                                                                              
  "error": {                                                                                                                                                   
    "code": "LinkedAuthorizationFailed",                                                                                                                       
    "message": "The client 'b96bfc14-d5f0-4982-baf7-094a9675db47' with object id 'b96bfc14-d5f0-4982-baf7-094a9675db47' has permission to perform action 'Micro
soft.Compute/virtualMachines/write' on scope '/subscriptions/0d202bbb-4fa7-4af8-8125-58c269a05435/resourceGroups/tomcontrast_caa_cluster/providers/Microsoft.Co
mpute/virtualMachines/podvm-openssl-backend-64f988dbd4-b562v-6e53dddd'; however, it does not have permission to perform action(s) 'Microsoft.Compute/galleries/
images/versions/read' on the linked scope(s) '/subscriptions/0d202bbb-4fa7-4af8-8125-58c269a05435/resourceGroups/tomcontrast_caa/providers/Microsoft.Compute/ga
lleries/contrast/images/contrast/versions/0.0.1' (respectively) or the linked scope(s) are invalid."                                                           
  }                                                                                                                                                            
}                                                                                                                                                              
--------------------------------------------------------------------------------                                                                               
```

Seems like you were right after all.

Comment (Freax13) [reply infra/azure-peerpods/vars.tf:13]: Hmm, not it's working again.

Comment (burgerdev) [reply infra/azure-peerpods/vars.tf:13]: I've seen this error, too, when working with peer pods, and giving explicit read access to the image resolves it. However, I also noticed that it does not appear to break things, which is kind of strange.

File: justfile
Comment (katexochen) [inline justfile:367]: This is in the wrong section, should go further down as it doesn't need to be changed.

File: justfile
Comment (Freax13) [reply justfile:323]: We do that because we want to destroy the resource group named `$azure_resource_group`. Tthis resource group contains the image gallery. If it doesn't exist yet, we create this resource group shortly before uploading the image: https://github.com/edgelesssys/contrast/blob/105ffe01ef0a2191849d6bdaf360b3b141e40f7f/packages/scripts.nix#L60-L63

File: justfile
Comment (katexochen) [inline justfile:321]:
```suggestion
            rm -f ${CONTRAST_CACHE_DIR}/image-upload/*.image-id
```
nit: Maybe we could have this in a separate directory, similar to the kds cache.

File: justfile
Comment (katexochen) [reply justfile:323]: I think we shouldn't share the aks resource group to also upload images there. I want to be able to destroy my aks setup without deleting the images that I use in a totally different peerpod cluster.

Also, I'd say it is better to inline both group creation and deletion directly into the justfile, so it's easier to understand what is happening. These are one-liners anyway.

File: packages/scripts.nix
Comment (katexochen) [inline packages/scripts.nix:32]:
```suggestion
```
set by the builder

File: packages/scripts.nix
Comment (katexochen) [inline packages/scripts.nix:58]: Debugging leftover or intended?

**Comment by katexochen** [comment]
I think the caching isn't correct the way it is currently implemented: If I upload `/nix/store/AAA-image-podvm` and then `/nix/store/BBB-image-podvm`, the cache will contain two files, each named after one of the nix path, both containing the same image reference. If I undo the change from BBB back to AAA, and want to redeploy, the upload will check the cache and find a file for the path AAA with the image reference and assumes it is already uploaded, but under that reference is actually the image from the BBB path.

File: packages/scripts.nix
Comment (katexochen) [inline packages/scripts.nix:34]: we should add location as var to the terraform to prevent location mismatch between cluster and image location.

File: packages/scripts.nix
Comment (Freax13) [reply packages/scripts.nix:58]: Neither, I copied parts of the script from `create-coco-aks.sh` and that line was already present there (it's also in `destroy-coco-aks.sh`). I'm not invested in having this, we can keep it or remove it, I'd be fine either way.

**Comment by Freax13** [comment]
> I think the caching isn't correct the way it is currently implemented: If I upload `/nix/store/AAA-image-podvm` and then `/nix/store/BBB-image-podvm`, the cache will contain two files, each named after one of the nix path, both containing the same image reference. If I undo the change from BBB back to AAA, and want to redeploy, the upload will check the cache and find a file for the path AAA with the image reference and assumes it is already uploaded, but under that reference is actually the image from the BBB path.

Good catch, I'll integrate the store hash into the image reference to avoid this.

File: justfile
Comment (Freax13) [reply justfile:323]: > I think we shouldn't share the aks resource group to also upload images there. I want to be able to destroy my aks setup without deleting the images that I use in a totally different peerpod cluster.

Do you want me to move the image gallery into the same resource group that contains the cluster or do you want the image gallery to be in another, new resource group?

**Comment by katexochen** [comment]
> Good catch, I'll integrate the store hash into the image reference to avoid this.

Not sure that's easily possible. :( Azure's image infra is crap. Image versions are restricted to semver, but having a separate gallery for each image requires additional time for creation during upload.

File: packages/scripts.nix
Comment (Freax13) [inline packages/scripts.nix:76]: Apparently Azure doesn't allow multiple galleries with the same name even if they are in different resource groups:

image

**Comment by Freax13** [comment]
> > Good catch, I'll integrate the store hash into the image reference to avoid this.
> 
> Not sure that's easily possible. :( Azure's image infra is crap. Image versions are restricted to semver, but having a separate gallery for each image requires additional time for creation during upload.

Turns out the limits on the version are pretty lax, so we can just use `0.0.<unix timestamp>` :)

File: justfile
Comment (katexochen) [reply justfile]: I'd say a separate group would be nice. If you aren't working on the image part, you might be happy with the same image for long time. On the other hand I'm not sure how to integrate that into the dev workflow, as it would require additional create/destroy steps for that. :thinking:

Comment (Freax13) [reply justfile]: That seems reasonable. Given that this PR has already dragged on for quite a while, are you fine with doing this in a follow-up PR?

**Comment by katexochen** [comment]
static checks need fixing

**Comment by Freax13** [comment]
@burgerdev @3u13r can one of you review this pr?

File: infra/azure-peerpods/vars.tf
Comment (burgerdev) [reply infra/azure-peerpods/vars.tf:13]: Why not just revive the `ra_image` above?

File: packages/scripts.nix
Comment (burgerdev) [inline packages/scripts.nix:37]: You could do `az account show -o tsv --query id`, which would avoid the hard-coded Edgeless subscription.

**Review by burgerdev — APPROVED** [review]

**Review by 3u13r — APPROVED** [review]

File: packages/scripts.nix
Comment (Freax13) [reply packages/scripts.nix:37]: This doesn't work if the user is part of multiple subscriptions though, does it?

Comment (burgerdev) [reply packages/scripts.nix:37]: It would show the default subscription. Maybe we should default to`az account show` but allow overriding with an explicit flag for the rare use case?

Comment (Freax13) [reply packages/scripts.nix:37]: We don't really hard-code this anyway, the value can be changed in `justfile.env`. We only hard-code the default value. Interpolating values into the just variable seems a bit painful, so let's just keep hard-coding the Edgeless subscription.

**Comment by Freax13** [comment]
I have two approvals, but Paul's last review requested changes, so I can't merge this. Given that Paul is on vacation until Wednesday, and has said that he's fine with merging this if there are approvals by other reviewers, I'll temporarily disable the merge requirement to merge this PR.

**Pull request merged**
